### ISSUE TITLE
About cabicgstab methods 

### ISSUE THREAD

**Issue by Rushsurf** [eccarson/ca-ksms#2]
Thanks a lot !This issue has been fixed ,the answer is right...
A few days ago,I replied to your answer,you may not see it,so I open another issue.If convient,can you leave your email address for communication?
And I have a question about CA-bicgstab method.I hope receive your reply.
I want to use CA-bicgstab methods with MPI parallel communication.However,I cant't find any communication statments in your code.
'CA' means 'communication avoiding'...
So,this code is written without any parallel communication,just serial?

**Comment by eccarson** [comment]
Yes, this is only MATLAB code meant to demonstrate the numerical behavior of the method. You can find a description of the parallel communication that would be needed in an MPI implementation of the method in this paper: https://epubs.siam.org/doi/10.1137/120881191

Namely, lines 66 and 69 in the code will require communication in performing the matrix-vector products. The particular MPI functions you should use will depend on the representation/partitioning of your operator and it's sparsity structure. 

Lines 80 and 83 will involve an MPI_Allreduce; the code can be implemented so that lines 80 and 83 are computed in a single reduction.

**Comment by Rushsurf** [comment]
[cabicgstab_example.txt](https://github.com/eccarson/ca-ksms/files/2559998/cabicgstab_example.txt)

I have finnished my code.However,the time (spent in iteratives) is not as good as i thought.
It's about the variable 's'.I opened 20 processes in my MPI parallel program,when 's' increases,the total iteration time does not decrease,but instead increase.
so i try the matlab code.
details as follows: 
n=2000 basis_type='monomial' 
s=1  time=0.4678s
s=2 time=0.6268s
s=3 time=0.6697s
s=4 time=0.9544s
as s increases,the tatal iteration time increases.....
I don't know why,because in my imagination,as 's' increases,the time spent will decrease to a certain extent,then increase ,or not converge.Right?
I hope that you can help me solve the doubts about 's' ,and about 'the time'  so much.
Thanks much!

**Comment by eccarson** [comment]
I do not expect the time to decrease with s in the MATLAB implementation. Without more details about your MPI implementation, it is difficult to say where the time is being spent there. Note that the MATLAB code includes things like computation of the true residual in each iteration to check for convergence, which should not be done in every iteration of an MPI implementation as it requires communication.

**Comment by Rushsurf** [comment]
Thank you very much for replying to me !
I have already taken your reminders into consideration.My purpose is to compare the two algorithems BICGSTAB and CA-BICGSTAB.
I have doubts about 's'. CA-BICGSTAB is improved from BICGSTAB.In a MPI parallel program,CA-bicgstab methods should spend less time compared to bicgstab methods (to some extent).
For example,
I opened 20 processes in my MPI parallel program.s=1,the total time and iteration steps,BICGSTAB and CA-BICGSTAB is almost the same.Howerver ,when 's' increases,the total 
time does not decrease,but instead increase.And i have a question about the iteration steps,  is “ one iteraton step after per s inner loop”？
And I have doubts about 's'.The value of 's' , any connection with total iteration time,iteration steps, and  convergence? I am not sure....
